backend/src/sms/services/twilio-client.class.ts

```diff
@@ -11,40 +13,53 @@ export default class TwilioClient {
     this.messagingServiceSid = messagingServiceSid
   }
 
-  public send(
+  async send(
     recipient: string,
     message: string,
     forceDelivery = false
   ): Promise<string | void> {
-    return this.client.messages
-      .create({
+    try {
+      const {
+        status,
+        sid,
+        error_code: errorCode,
+        code,
+      } = await this.client.messages.create({
         to: recipient,
         body: this.replaceNewLines(message),
         from: this.messagingServiceSid,
-        forceDelivery: forceDelivery,
+        forceDelivery,
       })
-      .then((result: { [key: string]: string }) => {
-        const { status, sid, error_code: errorCode, code } = result
-        if (sid) {
-          if (errorCode || code) {
-            return Promise.reject(new Error(`${sid};${errorCode};${code}`))
-          } else {
-            return sid
-          }
-        } else {
-          return Promise.reject(new Error(`${status};Unknown error`))
-        }
-      })
-      .catch((error: Error) => {
-        // Handle +65802 errors by forcing delivery once
-        if (
-          /\+65802\d+ is not a valid phone number/.test(error.message) &&
-          !forceDelivery
-        ) {
-          return this.send(recipient, message, true)
-        }
-        return Promise.reject(new Error(error.message))
+
+      if (!sid) {
+        throw new Error(`${status};Unknown error`)
+      }
+
+      if (errorCode || code) {
+        throw new Error(`${sid};${errorCode};${code}`)
+      }
+
+      return sid
+    } catch (error) {
+      // Handle +65802 errors by forcing delivery once
+      if (
+        /\+65802\d+ is not a valid phone number/.test(error.message) &&
+        !forceDelivery
+      ) {
+        return this.send(recipient, message, true)
+      }
+
+      // 20429 - REST API rate limit exceeded
+      // 21611 - Message queue limit exceeded
+      if ([20429, 21611].includes(error.code)) {
+        throw new RateLimitError()
+      }
+
+      throw new TwilioError({
```

Comment:
Do we really need to introduce `TwilioError`? `error.status` will be undefined for the error thrown on line 35 and 39.

Reply:
Ah my bad about `error.status` being undefined.

I was originally intending for the HTTP status code to vary based on the metadata of the TwilioError.

However, on second thought, there are too many error types to accurately map to HTTP status codes. And...the implementation is broken 😅

I think I'll just return a generic HTTP 500 status code along with the error message.

Reply:
Reverted https://github.com/opengovsg/postmangovsg/pull/1017/commits/100c283591a26128e7b8e423bdd9a09d40362e9d